File: Functions/run_BIOT.R
RunBIOT <- function(X, Fe, lambda, max.iter = 2, eps = 1e-6, rotation = F){
  # X: embedding matrix (response)
  # Fe: external feature matrix (predictors)
  # lambda: sparsity hyperparameter
  # max.iter: maximum number of iterations
  # eps: convergence threshold
  # rotation: should the orthogonal matrix be a rotation matrix? Yes = T, No = F

  Fe <- as.matrix(Fe)
  W <- matrix(1, nrow = ncol(Fe), ncol = ncol(X))
  R_squared <- matrix(1, nrow = ncol(X), ncol = 1)
  diff <- Inf
  iter <- 1
  crit <- list(Inf)
 
  while(iter < max.iter && diff > eps){
    # print(paste("Iter: ", iter))
    
    decomp <- svd( (1/(2*nrow(Fe))) * t(X) %*% Fe %*% W)
    # write.csv(decomp$u, file = "R_u.csv", row.names = FALSE)
    # write.csv(decomp$d, file = "R_d.csv", row.names = FALSE)
    # write.csv(decomp$v, file = "R_v.csv", row.names = FALSE)
    
    if (rotation == T){ # If rotation matrix is desired
      
      sv <- decomp$d
      smallest <- which.min(sv)
      sv[smallest] <- sign(det(decomp$u%*%t(decomp$v)))
      sv[-smallest] <- 1
      R <- (decomp$u%*%diag(sv)%*%t(decomp$v))
      
    } else { # If orthogonal matrix is desired (more general)
      
      R <- (decomp$u%*%t(decomp$v))
      
    }
    
    # W <- matrix(1, nrow = ncol(Fe), ncol = ncol(X))
    # R_squared <- matrix(1, nrow = ncol(X), ncol = 1)

    
    crit[[iter + 1]] <- BIOTCrit(Fe, X, R, W, lambda)
    diff <- abs(crit[[iter]] - crit[[iter + 1]])
    
    iter <- iter + 1
    
  }
  
  return(list(R = R, W = W, iter = iter, crit = unlist(crit), R_squared = R_squared))
  
}

BIOTCrit <- function(Fe, X, R, W, lambda){
  # Fe: external feature matrix (predictors)
  # X: embedding matrix (response)
  # R: orthogonal transformation matrix
  # W: regression weights
  # lambda: sparsity hyperparameter
  
  (1/(2*nrow(Fe)))*sum(diag(t(X - Fe%*%W%*%t(R))%*%((X - Fe%*%W%*%t(R))))) + lambda*sum(abs(W))
}

# # TESTING

# Read data from CSV file
X <- as.matrix(read.csv("X_norm_r.csv"))
Fe <- as.matrix(read.csv("Fe_norm_r.csv"))


K <- 30
for (i in 1:K) {
  s <- Sys.time()

  res <- RunBIOT(X, Fe, 0.0001, rotation = TRUE)

  elapsed <- Sys.time() - s
  print(elapsed)
}

# Print the iteration number
cat(sprintf("Iter: %d\n", res$iter))

# Save the matrices to CSV files
write.csv(res$R, "runBIOT/Rmx_R.csv", row.names = FALSE)
write.csv(res$W, "runBIOT/W_R.csv", row.names = FALSE)
write.csv(res$crit, "runBIOT/crit_R.csv", row.names = FALSE)
write.csv(res$r2, "runBIOT/r2_R.csv", row.names = FALSE)

File: PyFunctions/run_BIOT.py
import PyFunctions.get_W_Lasso as getW
import torch

def RunBIOT(X, Fe, lam, maxIter = 200, eps = 1e-6, rotation = False, device='cpu'):
  '''
  X: embedding matrix (response)
  Fe: external feature matrix (predictors)
  lambda: sparsity hyperparameter
  max.iter: maximum number of iterations
  eps: convergence threshold
  rotation: should the orthogonal matrix be a rotation matrix? Yes = T, No = F
  '''

  # Store all variables in the appropriate device memory, ideally, GPU if available.
  W, r2 = getW.Lasso(Fe, X, lam, device=device)
  n = Fe.shape[0]

  diff = 1
  iter = 0
  crit = []
  crit.append(1)

  while( iter + 1 < maxIter and diff > eps ):  
    # Perform singular value decomposition
    u, d, v = torch.linalg.svd( (1 / (2 * n)) * X.t() @ (Fe @ W))

    if rotation:
      # Set the smallest singular value to the sign of the determinant of U and V^T.
      # Set everything else to equal 1.
      smallest = torch.argmin(d)
      d[smallest] = torch.sign(torch.linalg.det(u @ v.T))
      d[d != d[smallest]] = 1

      R = u @ torch.diag(d) @ v.T

    else:
      # If rotation is not preferred, do not multiply R times the diagonal of d.
      R = u @ v.t()
    
    # Find the new weights and r^2 
    W, r2 = getW.Lasso(Fe, X@R, lam, device=device)

    # Find the difference between the critical values of the current and previous iteration.
    # We will stop iterating once this difference is less than epsilon.
    crit.insert(iter + 1, BIOTCrit(Fe, X, R, W, lam))
    diff = abs(crit[iter] - crit[iter + 1])

    # On to the next iteration. 
    iter += 1
  

  return ( R, W, iter, crit, r2 ) 



def BIOTCrit(Fe, X, R, W, lam):
  '''  
  Fe: external feature matrix (predictors)
  X: embedding matrix (response)
  R: orthogonal transformation matrix
  W: regression weights
  lambda: sparsity hyperparameter
  '''

  n = 1 / ( 2 * Fe.shape[0] )
  mx = X - Fe @ W @ R.T
  loss = lam * torch.sum( abs(W) )

  return ( n * torch.sum( torch.diag( mx.T @ mx)) + loss )


# TEST
# X = torch.tensor(np.genfromtxt("X_train_norm_py.csv", delimiter=',', dtype='float64'))
# Fe = torch.tensor(np.genfromtxt("Fe_train_norm_py.csv", delimiter=',', dtype='float64'))
# R, W, crit, iter, r2 = RunBIOT(X, Fe,  0.0001, rotation=True)

# Pu = np.genfromtxt("Py_svd.csv", delimiter=',', dtype='float64')
# Ru = np.genfromtxt("R_svd.csv", delimiter=',', skip_header=1, dtype='float64')
# # diff = abs(Ru - Pu)
# # print(diff[diff > 1e-2])
# np.savetxt('Rsvd_Psvd.csv', Ru - Pu, delimiter=',')

# Pu = np.genfromtxt("Py_u.csv", delimiter=',', dtype='float64')
# Ru = np.genfromtxt("R_u.csv", delimiter=',', skip_header=1, dtype='float64')
# np.savetxt('Ru_Pu.csv', Ru - Pu, delimiter=',')

# Pu = np.genfromtxt("Py_d.csv", delimiter=',', dtype='float64')
# Ru = np.genfromtxt("R_d.csv", delimiter=',', skip_header=1, dtype='float64')
# print( Ru - Pu )

# Pu = np.genfromtxt("Py_v.csv", delimiter=',', dtype='float64')
# Ru = np.genfromtxt("R_v.csv", delimiter=',', skip_header=1, dtype='float64')
# print( Ru - Pu )

if __name__ == "__main__":
    import numpy as np
    import time
    import tracemalloc

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print(f"Device: {device}\n")
    home = ""
    X = torch.tensor(np.genfromtxt(home + "X_norm_r.csv", delimiter=',', dtype='float64'))
    Fe = torch.tensor(np.genfromtxt(home + "Fe_norm_r.csv", delimiter=',', dtype='float64'))        
    lam = torch.tensor([0.0001], dtype=torch.float64, device=device)
    
    def testing():
        K = 30
        times = []
        for i in range(K):

            s = time.time()

            R, W, crit, iter, r2 = RunBIOT(X, Fe,  lam, rotation=True, device=device)

            times.append(time.time() - s)
        np.savetxt(f"run_BIOT_times_py.csv", times, delimiter=",")
        print(f"Iter: {iter}\n")

        np.savetxt("R_py.csv", R, delimiter=",")
        np.savetxt("W_py.csv", W, delimiter=",")
        np.savetxt("crit_py.csv", crit, delimiter=",")
        np.savetxt("r2_py.csv", r2, delimiter=",")


    def memtrace():
        tracemalloc.start()
        snapshot_before = tracemalloc.take_snapshot()

        R, W, crit, iter, r2 = RunBIOT(X, Fe,  lam, rotation=True, device=device)

        snapshot_after = tracemalloc.take_snapshot()

        difference = snapshot_after.compare_to(snapshot_before, 'lineno')
        mems = []
        for stat in difference:
            mems.append(stat)
        print(difference)
        np.savetxt(f"run_BIOT_mems_py.csv", mems, delimiter=",")

    testing()
    memtrace()